File: sdk-python/obsvr/govern_fn.py
"""Ergonomic governance for application-owned callables."""

from __future__ import annotations

import time
from typing import Any, Callable, Optional, TypeVar, overload

from .binding_report import record_binding
from .integrations.tools import govern_tool

F = TypeVar("F", bound=Callable[..., Any])
_MAX_TEXT_BYTES = 256
_GOVERNED_FUNCTION_MARKER = "__obsvr_governed_function__"


def _bounded_text(value: Any, field: str, *, required: bool = False) -> Optional[str]:
    if value is None and not required:
        return None
    if not isinstance(value, str) or not value.strip():
        raise TypeError(f"{field} must be a nonblank string")
    normalized = value.strip()
    if len(normalized.encode("utf-8")) > _MAX_TEXT_BYTES:
        raise TypeError(f"{field} exceeds {_MAX_TEXT_BYTES} UTF-8 bytes")
    return normalized


def govern_fn(
    fn: F,
    *,
    name: Optional[str] = None,
    surface: str = "action",
    consequence: Optional[str] = None,
    description: Optional[str] = None,
    user_id: Optional[str] = None,
    service_name: Optional[str] = None,
    metadata: Optional[dict[str, Any]] = None,
) -> F:
    """Return a governed callable with the same sync or async shape as ``fn``.

    The implementation reuses :func:`govern_tool`, so block happens before the
    function body and redaction changes the arguments the function receives.
    A retained reference to the original function remains a bypass and is
    recorded as a coverage exclusion.
    """
    if not callable(fn):
        raise TypeError("fn must be callable")
    if getattr(fn, _GOVERNED_FUNCTION_MARKER, False) is True:
        return fn

    resolved_name = _bounded_text(
        name if name is not None else getattr(fn, "__name__", None),
        "name",
        required=True,
    )
    if surface not in {"action", "workflow"}:
        raise TypeError("surface must be action or workflow")
    resolved_consequence = _bounded_text(consequence, "consequence")
    resolved_description = _bounded_text(description, "description")
    action_metadata = {
        "surface": surface,
        "name": resolved_name,
        **(
            {"consequence": resolved_consequence}
            if resolved_consequence is not None
            else {}
        ),
    }
    governed = govern_tool(
        fn,
        name=resolved_name,
        descriptor={
            "name": resolved_name,
            **(
                {"description": resolved_description}
                if resolved_description is not None
                else {}
            ),
        },
        user_id=user_id,
        service_name=service_name,
        metadata={**(metadata or {}), "obsvr_action": action_metadata},
    )
    setattr(governed, _GOVERNED_FUNCTION_MARKER, True)
    record_binding(
        "govern_fn",
        resolved_name,
        metadata={
            "enforcement_depth": "enforce",
            "initialized_at_ms": int(time.time() * 1000),
            "exclusions": ["calls through retained raw function aliases"],
        },
    )
    return governed


@overload
def govern(fn: F, /) -> F: ...


@overload
def govern(
    fn: None = None,
    /,
    **options: Any,
) -> Callable[[F], F]: ...


def govern(fn: Optional[F] = None, /, **options: Any) -> Any:
    """Decorator or direct wrapper form of :func:`govern_fn`."""

    def decorate(target: F) -> F:
        return govern_fn(target, **options)

    return decorate if fn is None else decorate(fn)


__all__ = ["govern", "govern_fn"]

File: sdk-typescript/src/governance/govern-fn.ts
import { recordBinding } from '../binding-report.js';
import {
  obsvrGovernTool,
  type GovernToolOptions,
} from '../integrations/tools.js';

const GOVERNED_FUNCTION_MARKER = Symbol.for('obsvr.governedFunction');
const MAX_ACTION_NAME_BYTES = 256;

type AnyFunction = (...args: never[]) => unknown;

export type GovernedFunction<F extends AnyFunction> = (
  ...args: Parameters<F>
) => Promise<Awaited<ReturnType<F>>>;

export interface GovernFnOptions extends GovernToolOptions {
  /** Stable action identifier used by policy, evidence, and coverage reports. */
  name?: string;
  /** Application boundary represented by the callable. */
  surface?: 'action' | 'workflow';
  /** Operator-defined consequence class, for example `external_write`. */
  consequence?: string;
  /** Human-readable descriptor sealed with the governed tool descriptor. */
  description?: string;
}

function boundedText(value: unknown, field: string, required = false): string | undefined {
  if (value === undefined && !required) return undefined;
  if (typeof value !== 'string' || value.trim() === '') {
    throw new TypeError(`${field} must be a nonblank string`);
  }
  const normalized = value.trim();
  if (Buffer.byteLength(normalized, 'utf8') > MAX_ACTION_NAME_BYTES) {
    throw new TypeError(`${field} exceeds ${MAX_ACTION_NAME_BYTES} UTF-8 bytes`);
  }
  return normalized;
}

/**
 * Govern any application callable at its invocation boundary.
 *
 * The returned function is always async because it reuses the complete
 * pre-call tool pipeline, including approval waits and external policy. A deny
 * rejects before `fn` is entered. A redact verdict rewrites the arguments the
 * function receives, or fails closed when the rewrite cannot be proven.
 *
 * Keeping another reference to `fn` is a bypass. The coverage binding records
 * that exclusion rather than presenting this wrapper as process-wide control.
 */
export function governFn<F extends AnyFunction>(
  fn: F,
  options: GovernFnOptions = {},
): GovernedFunction<F> {
  if (typeof fn !== 'function') throw new TypeError('fn must be callable');
  const marked = fn as F & { [GOVERNED_FUNCTION_MARKER]?: boolean };
  if (marked[GOVERNED_FUNCTION_MARKER] === true) {
    return fn as unknown as GovernedFunction<F>;
  }

  const name = boundedText(options.name ?? fn.name, 'name', true) as string;
  const surface = options.surface ?? 'action';
  if (surface !== 'action' && surface !== 'workflow') {
    throw new TypeError('surface must be action or workflow');
  }
  const consequence = boundedText(options.consequence, 'consequence');
  const description = boundedText(options.description, 'description');
  const actionMetadata = Object.freeze({
    surface,
    name,
    ...(consequence === undefined ? {} : { consequence }),
  });

  const governed = obsvrGovernTool({
    name,
    ...(description === undefined ? {} : { description }),
    execute: async (payload: { args: Parameters<F> }) => fn(...payload.args),
  }, {
    source: options.source ?? 'obsvr_govern_fn',
    region: options.region,
    service_name: options.service_name,
    user_id: options.user_id,
    metadata: {
      ...(options.metadata ?? {}),
      obsvr_action: actionMetadata,
    },
    name,
  });

  const wrapped = (async (...args: Parameters<F>) => governed.execute({ args })) as GovernedFunction<F> & {
    [GOVERNED_FUNCTION_MARKER]?: boolean;
  };
  Object.defineProperty(wrapped, GOVERNED_FUNCTION_MARKER, {
    value: true,
    enumerable: false,
    configurable: false,
  });
  Object.defineProperty(wrapped, 'name', {
    value: fn.name || name,
    configurable: true,
  });

  recordBinding('govern_fn', name, undefined, {
    enforcementDepth: 'enforce',
    initializedAtMs: Date.now(),
    exclusions: ['calls through retained raw function aliases'],
  });
  return wrapped;
}
